File: hooks/cite-fidelity-lint.py
#!/usr/bin/env -S uv run python3
"""PostToolUse hook: lint a freshly-written drafts/*.md for cite-fidelity issues.

Fires after Edit|Write of a markdown file inside the writing project's
drafts/ directory. Runs Stage 4 lint (handcite + bundled + bibkey-not-in-nlm
unless ACTIVE_WORKFLOW.md has no notebook) and surfaces findings as a
non-blocking stderr warning visible in the conversation.

Always defaults to approve — never blocks tool calls. The hook's job is to
nudge the agent toward Stage 2 grounding, not to refuse edits.

Skipped when:
- No file_path in tool_input
- Edited file is not under drafts/
- ACTIVE_WORKFLOW.md is unreachable (cite-fidelity not configured)
- Any unexpected error (defensive default-approve)
"""
from __future__ import annotations

import json
import subprocess
import sys
from pathlib import Path


def main() -> int:
    try:
        hook_input = json.load(sys.stdin)
    except Exception:
        return 0

    tool_name = hook_input.get("tool_name", "")
    if tool_name not in ("Edit", "Write", "MultiEdit"):
        return 0

    tool_input = hook_input.get("tool_input", {}) or {}
    file_path = tool_input.get("file_path", "")
    if not file_path:
        return 0

    fp = Path(file_path)
    if fp.suffix != ".md":
        return 0
    if "drafts" not in fp.parts:
        return 0

    plugin_root = Path(__file__).resolve().parent.parent
    lint_script = plugin_root / "scripts" / "cite-fidelity" / "lint_drafts.py"
    if not lint_script.exists():
        return 0

    project_dir = fp.parent
    while project_dir != project_dir.parent:
        if (project_dir / ".planning" / "ACTIVE_WORKFLOW.md").exists():
            break
        project_dir = project_dir.parent
    else:
        return 0

    try:
        result = subprocess.run(
            ["uv", "run", "python3", str(lint_script), str(fp)],
            capture_output=True, text=True, timeout=60,
            cwd=str(project_dir),
        )
    except Exception:
        return 0

    out = (result.stdout or "").strip()
    err = (result.stderr or "").strip()
    if not out and not err:
        return 0
    if "✓ no issues" in out:
        return 0

    lines = ["[cite-fidelity-lint] findings on " + fp.name + ":"]
    if out:
        lines.append(out)
    if err:
        lines.append(err)
    payload = {
        "hookSpecificOutput": {
            "hookEventName": "PostToolUse",
            "additionalContext": "\n".join(lines),
        }
    }
    print(json.dumps(payload))
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: hooks/cite-fidelity-lint.ts
#!/usr/bin/env bun
/** PostToolUse hook: lint a freshly-written drafts/*.md for cite-fidelity issues.
 *
 * TypeScript port of cite-fidelity-lint.py — behavior-identical, including the odd bits.
 *
 * Fires after Edit|Write|MultiEdit of a markdown file inside the writing project's drafts/
 * directory. Runs Stage 4 lint and surfaces findings as non-blocking additionalContext.
 *
 * Always defaults to approve — never blocks tool calls, never exits non-zero.
 *
 * Skipped when: no file_path, not a .md, not under a `drafts` path component, the bundled lint
 * script is missing, no receipt-selected canonical plan state above the file, or any unexpected error.
 *
 * PATH SEMANTICS: this hook uses pathlib-like `suffix`, `parts`, and `parent` behavior while walking
 * upward from a draft. A project qualifies only when its hidden review receipt selects and
 * authenticates a current approved writing plan.
 */
import { existsSync, realpathSync } from "node:fs";
import { context, readPayload } from "./_gate_common.ts";
import { authenticatedWritingPlan } from "./lib/writing-plan-context.ts";

/** PurePosixPath-alike: just enough of pathlib for this hook. */
class PPath {
  readonly root: string;
  readonly parts: string[];
  constructor(root: string, parts: string[]) {
    this.root = root;
    this.parts = parts;
  }
  static parse(p: string): PPath {
    return new PPath(p.startsWith("/") ? "/" : "", p.split("/").filter((x) => x !== "" && x !== "."));
  }
  /** str(Path(...)): "." for the empty relative path. */
  get str(): string {
    if (this.parts.length === 0) return this.root || ".";
    return this.root + this.parts.join("/");
  }
  /** pathlib's `.parts`: absolute paths carry a leading "/" element. */
  get pathParts(): string[] {
    return this.root ? [this.root, ...this.parts] : [...this.parts];
  }
  get name(): string {
    return this.parts.length ? this.parts[this.parts.length - 1] : "";
  }
  /** pathlib: no suffix for a leading dot (".bashrc") or a trailing one ("a."). */
  get suffix(): string {
    const n = this.name;
    const i = n.lastIndexOf(".");
    return i > 0 && i < n.length - 1 ? n.slice(i) : "";
  }
  get parent(): PPath {
    if (this.parts.length === 0) return this;
    return new PPath(this.root, this.parts.slice(0, -1));
  }
  join(...segs: string[]): string {
    const base = this.str;
    return base === "/" ? "/" + segs.join("/") : base + "/" + segs.join("/");
  }
}

async function main(): Promise<number> {
  let hookInput: Record<string, unknown>;
  try {
    hookInput = await readPayload();
  } catch {
    return 0;
  }

  const toolName = String(hookInput?.tool_name ?? "");
  if (toolName !== "Edit" && toolName !== "Write" && toolName !== "MultiEdit") return 0;

  const toolInput = (hookInput?.tool_input as Record<string, unknown>) || {};
  const filePath = String(toolInput?.file_path ?? "");
  if (!filePath) return 0;

  const fp = PPath.parse(filePath);
  if (fp.suffix !== ".md") return 0;
  if (!fp.pathParts.includes("drafts")) return 0;

  const pluginRoot = PPath.parse(realpathSync(import.meta.path)).parent.parent;
  const lintScript = pluginRoot.join("scripts", "cite-fidelity", "lint_drafts.py");
  if (!existsSync(lintScript)) return 0;

  const plan = authenticatedWritingPlan(fp.parent.str);
  if (!plan) return 0;

  let stdout = "";
  let stderr = "";
  try {
    const result = Bun.spawnSync({
      cmd: ["uv", "run", "python3", lintScript, "--project-root", plan.projectRoot, ...(plan.notebook ? ["--nlm-notebook", plan.notebook] : []), fp.str],
      cwd: plan.projectRoot,
      stdout: "pipe",
      stderr: "pipe",
      timeout: 60_000,
    });
    // subprocess.run(timeout=60) RAISES on timeout, so the Python swallows it and returns 0.
    if ((result as { exitedDueToTimeout?: boolean }).exitedDueToTimeout) return 0;
    stdout = result.stdout ? result.stdout.toString() : "";
    stderr = result.stderr ? result.stderr.toString() : "";
  } catch {
    return 0;
  }

  const out = stdout.trim();
  const err = stderr.trim();
  if (!out && !err) return 0;
  if (out.includes("✓ no issues")) return 0;

  const lines = ["[cite-fidelity-lint] findings on " + fp.name + ":"];
  if (out) lines.push(out);
  if (err) lines.push(err);
  context("PostToolUse", lines.join("\n")); // never returns
}

process.exit(await main());
